docs/hot_module_reloading.md

```diff
@@ -0,0 +1,190 @@
+out by dart2js as unused.
+
+```dart
+import 'dart:html';
+
+var _id = 'hello';
+
+void main() {
+  var helloDiv = Element.div()
+    ..id = _id
+    ..text = 'Hello Dart';
+  document.body.append(helloDiv);
+}
+
+Object hot$onDestroy() {
+  document.body.querySelectorAll('#$_id').forEach((e) => e.remove());
+}
+```
+
+### Handling reloading of child modules
+
+Lets assume you have a builder that transforms your css files into dart code exporting the string.
+As this string doesn't have any impact on your logic, you want to handle reloading of this modules,
+to prevent parent from reloading
+
+For simplicity of example, lets assume we have `addCss` and `removeCss` methods thad do real DOM
+modifications. In this example it will just operate with `Set`.
+
+```dart
+// your_package/lib/styles.dart
+
+var _styles = new Set<String>();
+
+void addCss(String css) {
+  _styles.add(css);
+}
+
+void removeCss(String css) {
+  _styles.remove(css);
+}
+```
+
+```dart
+// your_package/web/main.dart
+
+import 'package:js/js.dart';
+import 'package:your_package/src/css_file.css.shim.dart' as css_file;
+import 'package:your_package/styles.dart';
+
+// You should encapsulate usages of your library to one place you are able to substitute 
+// (non-final variable in this case)
+var _css_style = css_file.styles;
+
+void main() {
+  addCss(_css_style);
+  
+  // Some stateful code you don't want to reinitialise
+  // ...
+}
+
+// We need to describe library structure as js object to be able to use it
+@JS()
+abstract class StylesLib {
+  @JS()
+  external String get styles;
+}
+
+bool hot$onChildUpdate(String id, Object child) {
+  if (id == 'package:your_package/src/css_file.css.shim.dart') {
+    removeCss(_css_style);
+    _css_style = (child as StylesLib).styles;
+    addCss(_css_style);
+    return true;
+  }
+}
+```
+
+Another implementation of `hot$onChildUpdate` using `dart:js_util` may be the following:
+```dart
+import 'dart:js_util';
+
+bool hot$onChildUpdate(String id, Object child) {
+  if (id == 'package:your_package/src/css_file.css.shim.dart') {
+    removeCss(_css_style);
+    _css_style = getProperty(child, 'styles');
+    addCss(_css_style);
+    return true;
+  }
+}
+```
+
+## Known issues
+
+- Creating new modules, removing them or otherwise changing dependency grapth results in full page
+  reload. [#1761](https://github.com/dart-lang/build/issues/1761)
+- Libraries are often bundled together in one module. In most cases `hot$onChildUpdate` hook for
+  such bundled modules won't work - code will be executed, but parent module will still be reloaded.
+  That happens because current requirement is for all libraries in module to know how to handle
+  child updates. If you actively use this hook, you may consider turning on `fine` build strategy in
+  your `build.yaml`, work around this issue. But this will also slow down your builds. [#1767](https://github.com/dart-lang/build/issues/1767)
+  
+  ```yaml
+  global_options:
```

Comment:
Most people probably only need to actually enable this for the root package, I think it is probably worth showing how to do both. The config for a regular package is like this:

```yaml
targets:
  $default:
    builders:
      build_modules|modules:
        options:
          strategy: fine
```